http/src/main/scala/org/http4s/blaze/http/http2/WriteControllerImpl.scala

```diff
@@ -0,0 +1,154 @@
+  private[this] val interestedStreams = new java.util.ArrayDeque[WriteInterest]
+  private[this] val pendingWrites = new util.ArrayDeque[Seq[ByteBuffer]]
+
+  private[this] var state: State = Idle
+
+  def close(): Future[Unit] = state match {
+    case Idle =>
+      state = Closed
+      Future.successful(())
+
+    case Flushing =>
+      val p = Promise[Unit]
+      state = Closing(p)
+      p.future
+
+    case Closing(p) =>
+      p.future
+
+    case Closed =>
+      Future.successful(())
+  }
+
+  private[this] def pendingInterests: Boolean = !pendingWrites.isEmpty || !interestedStreams.isEmpty
+
+  def write(data: Seq[ByteBuffer]): Boolean = state match {
+    case Idle | Flushing | Closing(_) =>
+      pendingWrites.addLast(data)
+      maybeWrite()
+      true
+
+    case Closed =>
+      false
+  }
+
+  def write(data: ByteBuffer): Boolean = write(data::Nil)
+
+  def registerWriteInterest(interest: WriteInterest): Boolean = state match {
+    case Closed => false
+    case _ =>
+      interestedStreams.add(interest)
+      maybeWrite()
+      true
+  }
+
+  private[this] def maybeWrite(): Unit = {
+    if (state == Idle) {
+      state = Flushing
+      doWrite()
+    }
+  }
+
+  private[this] def addDirectWrites(dest: ArrayBuffer[ByteBuffer]): Int = {
+    var written = 0
+    while (!pendingWrites.isEmpty) {
+      written += addBuffs(dest, pendingWrites.poll())
+    }
+    written
+  }
+
+  private[this] def addBuffs(dest: ArrayBuffer[ByteBuffer], data: Seq[ByteBuffer]): Int = {
+    var written = 0
+    data.foreach { buf =>
+      val rem = buf.remaining
+      if (0 < rem) {
+        written += rem
+        dest += buf
+      }
+    }
+    written
+  }
+
+  // The meat and potatoes
+  private[this] def doWrite(): Unit = {
+    val toWrite = new ArrayBuffer[ByteBuffer]()
+    var bytesToWrite = addDirectWrites(toWrite)
+
+    // Accumulate bytes until we run out of interests or have exceeded the high-water mark
+    while(!interestedStreams.isEmpty && bytesToWrite < highWaterMark) {
+      try {
+        val data = interestedStreams.poll().performStreamWrite()
+        bytesToWrite += addBuffs(toWrite, data)
+      } catch { case NonFatal(t) =>
+        logger.error(t)(s"Unhandled exception performing stream write operation")
+      }
+    }
+
+    logger.debug(s"Flushing $bytesToWrite to the wire")
+
+    tailStage.channelWrite(toWrite).onComplete {
+      case Success(_) =>
+        state match {
+          case Idle =>
+            throw new IllegalStateException("Write finished to find Idle state")
+
+          case Flushing if pendingInterests =>
+            doWrite()
+
+          case Flushing =>
+            state = Idle
+
+          case Closing(_) if pendingInterests =>
+            doWrite()
+
+          case Closing(p) =>
+            state = Closed
+            p.trySuccess(())
+
+          case Closed =>
+            () // shouldn't get here
```

Comment:
If you shouldn't and did, you should log it.

Reply:
I'll throw an exception instead since there is a serious bug if we get here. The sessions `serialExecutor` will tear everything down on unhandled exceptions and make a fuss.